Add default tag to api conventions (#5723)
kubernetes/community: contributors/devel/sig-architecture/api-conventions.md

full desired state so that it is easier for the system to determine how to
achieve the state, and for the user to know what to anticipate.

Default values can be specified on a field using the `+default=` tag. Primitives
will have their values directly assigned while structs will go through the

> JSON unmarshalling process.

Reviewer: Maybe include an example of a semi-complex json form. Is there something specific in the kubebuilder book about the format of these that can be interesting? Also we need to mention that non-pointer fields can't default to anything but the zero value.

Author: Updated to include the non-pointer comment. I think for the json example, it would be more suited in the website documentation rather than here?

full desired state so that it is easier for the system to determine how to
achieve the state, and for the user to know what to anticipate.

Default values can be specified on a field using the `+default=` tag. Primitives

> will have their values directly assigned while structs will go through the

Reviewer 2: I thought we never converged on the struct default via JSON thing, did we already merge that?

Reviewer: https://github.com/kubernetes/kube-openapi/blob/master/test/integration/testdata/defaults/default.go#L11-L12

Reviewer 2: ok, great.

> JSON unmarshalling process. Fields that do not have an `omitempty` json tag may
> only default to the zero value of their corresponding type.

Reviewer 2: ```suggestion
JSON unmarshalling process. Fields that do not have an `omitempty` json tag will
default to the zero value of their corresponding type if no default is assigned.
```

Author: On a second pass, this doesn't seem to be correct. I'd like to say that fields that do not have an omitempty tag will default to the zero value regardless of whether a default is set or not. If the default is set, it must be set to the zero value or else an error will occur.

Reviewer 2: Why would we make it work that way? "omitempty" is meaningless for fields with a default.

Author: Oops sorry I missed this earlier. My explanation was a bit unclear. 

Yes, omitempty is meaningless when default is set, but there is a restriction of what the default value can be.

With the omitempty tag, the default value can be any value (within the domain of the type). Without the omitempty tag, the default value MUST be set to the to the zero value. This is because go structs for fields that are not pointers will automatically set the zero value for the field.

I still think the original explanation is the most accurate, although maybe a bit convoluted.

> default to the zero value of their corresponding type if no default is assigned.

Reviewer 2: This sentence is a double negative now :)
